src/Compilers/Core/Portable/SourceGeneration/GeneratorContexts.cs

```diff
@@ -70,8 +70,16 @@ internal GeneratorExecutionContext(Compilation compilation, ParseOptions parseOp
         /// Adds source code in the form of a <see cref="string"/> to the compilation.
         /// </summary>
         /// <param name="hintName">An identifier that can be used to reference this source text, must be unique within this generator</param>
-        /// <param name="source">The source code to be add to the compilation</param>
-        public void AddSource(string hintName, string source) => AddSource(hintName, SourceText.From(source, Encoding.UTF8));
+        /// <param name="source">The source code to add to the compilation</param>
+        public void AddSource(string hintName, string source) => AddSource(hintName, SourceText.From(source, Encoding.UTF8, SourceHashAlgorithm.Sha256));
+
+        /// <summary>
+        /// Adds source code in the form of a <see cref="StringBuilder"/> to the compilation.
+        /// This builder must not be further modified: undefined behavior will result.
+        /// </summary>
+        /// <param name="hintName">An identifier that can be used to reference this source text, must be unique within this generator</param>
+        /// <param name="source">The source code to add to the compilation. This builder must not be further modified: undefined behavior will result.</param>
```

Comment:
>must not be further modified [](start = 89, length = 28)

What happens if I get the `StringBuilder` from a `StringBuilderCache` becuase I'm trying to be more efficient?

Reply:
Yeah this seems like a particularly dangerous problem here that I would want to guard against.

Reply:
In general the policy is to not pass around mutable state between the compiler and plugins that would impact the output of the build. I hadn't caught on in first review that we don't immediately realize the contents of the `StringBuilder` here.